haskell/protobuf.bzl

```diff
@@ -0,0 +1,344 @@
+  """Camel-case the input string, preserving any existing capital letters.
+  """
+  # Split on both "-" and "_", matching the behavior of proto-lens-protoc.
+  return "".join([_capitalize_first_letter(c2)
+                  for c1 in comp.split("_") for c2 in c1.split("-")])
+
+def _proto_lens_output_file(path):
+  """The output file from `proto-lens-protoc` when run on the given `path`.
+  """
+
+  path = path[:-len(".proto")]
+  result = "/".join([_camel_case(p) for p in path.split("/")]) + ".hs"
+
+  return "Proto/" + result
+
+def _haskell_proto_aspect_impl(target, ctx):
+
+  args = ctx.actions.args()
+
+  src_prefix = paths.join(
+    ctx.label.workspace_root,
+    ctx.label.package,
+  )
+
+  args.add(["--plugin=protoc-gen-haskell=" +
+            protobuf_tools(ctx).plugin.path,
+  ])
+
+  hs_files = []
+  inputs = []
+
+  args.add(["-I{0}={1}".format(s.short_path, s.path)
+            for s in target.proto.transitive_sources])
+
+  inputs.extend(target.proto.transitive_sources.to_list())
+
+  for src in target.proto.direct_sources:
+
+    inputs.append(src)
+
+    # As with the native rules, require the .proto file to be in the same
+    # Bazel package as the proto_library rule. This allows us to put the
+    # output .hs file next to the input .proto file. Unfortunately Skylark
+    # doesn't let us check the package of the file directly, so instead we
+    # just look at its short_path and rely on the proto_library rule itself
+    # to check for consistency. We use the file's short_path rather than its
+    # dirname/basename in case it's in a subdirectory; for example, if the
+    # proto_library rule is in "foo/BUILD" but the .proto file is
+    # "foo/bar/baz.proto".
+
+    if not src.short_path.startswith(src_prefix):
+      fail("Mismatch between rule context " + str(ctx.label.package)
+           + " and source file " + src.short_path)
+    if src.basename[-6:] != ".proto":
+      fail("bad extension for proto file " + src)
+
+    relative_path = src.short_path
+
+    args.add([relative_path])
+
+    hs_files.append(ctx.actions.declare_file(
+      _proto_lens_output_file(relative_path),
+    ))
+
+  args.add([
+    "--haskell_out=no-reexports:" + paths.join(
+      hs_files[0].root.path,
+      src_prefix,
+    ),
+  ])
+
+  ctx.actions.run(
+    inputs = depset([
+      protobuf_tools(ctx).protoc,
+      protobuf_tools(ctx).plugin,
+    ] + inputs),
+    outputs = hs_files,
+    executable = protobuf_tools(ctx).protoc,
+    arguments = [args],
+  )
+
+  patched_attrs = dict(
+    {
+      "compiler_flags": [],
+      "src_strip_prefix": "",
+      "prebuilt_dependencies": [],
+      "repl_interpreted": True,
+      "repl_ghci_args": [],
+      "version": "1.0.0",
+      "_ghc_defs_cleanup": ctx.attr._ghc_defs_cleanup,
+      "_ghci_script": ctx.attr._ghci_script,
+      "_ghci_repl_wrapper": ctx.attr._ghci_repl_wrapper,
+      "hidden_modules": [],
+    },
+    name = "proto-autogen-" + ctx.rule.attr.name,
```

Comment:
Why is this using both the positional and the string-based ways of initializing the dictionary?  (For example, `"compiler_flags":...` vs `srcs =...`)